io-engine/src/rebuild/rebuilders.rs

```diff
@@ -0,0 +1,255 @@
+    /// Fetch the next block to rebuild.
+    fn next(&mut self) -> Option<u64>;
+    /// Peek the next block to rebuild.
+    fn peek_next(&self) -> Option<u64>;
+    /// Get the remaining blocks we have yet to be rebuilt.
+    fn blocks_remaining(&self) -> u64;
+    /// Check if this is a partial rebuild.
+    fn is_partial(&self) -> bool;
+    /// Get the rebuild descriptor reference.
+    fn desc(&self) -> &RebuildDescriptor;
+    /// Get the copier which can copy a segment.
+    fn copier(&self) -> Rc<T>;
+}
+
+/// The range is the full range of the request, in steps of segment size.
+pub(super) struct FullRebuild<T: RebuildTaskCopier> {
+    range: PeekableIterator<std::iter::StepBy<Range<u64>>>,
+    copier: Rc<T>,
+}
+impl<T: RebuildTaskCopier> FullRebuild<T> {
+    /// Create a full rebuild with the given copier.
+    pub(super) fn new(copier: T) -> Self {
+        let desc = copier.descriptor();
+        let range = desc.range.clone();
+        Self {
+            range: PeekableIterator::new(
+                range.step_by(desc.segment_size_blks as usize),
+            ),
+            copier: Rc::new(copier),
+        }
+    }
+}
+impl<T: RebuildTaskCopier> RangeRebuilder<T> for FullRebuild<T> {
+    fn next(&mut self) -> Option<u64> {
+        self.range.next()
+    }
+    fn peek_next(&self) -> Option<u64> {
+        self.range.peek().cloned()
+    }
+
+    fn blocks_remaining(&self) -> u64 {
+        self.peek_next()
+            .map(|r| self.desc().range.end.max(r) - r)
+            .unwrap_or_default()
+    }
+    fn is_partial(&self) -> bool {
+        false
+    }
+
+    fn desc(&self) -> &RebuildDescriptor {
+        self.copier.descriptor()
+    }
+    fn copier(&self) -> Rc<T> {
+        self.copier.clone()
+    }
+}
+
+/// A partial rebuild range which steps through each segment but triggers
+/// the copy only if the segment dirty bit is set.
+pub(super) struct PartialRebuild<T: RebuildTaskCopier> {
+    range: PeekableIterator<std::iter::Enumerate<bit_vec::IntoIter>>,
+    segment_size: u64,
+    total_blks: u64,
+    rebuilt_blks: u64,
+    copier: Rc<T>,
+}
+impl<T: RebuildTaskCopier> PartialRebuild<T> {
+    /// Create a partial sequential rebuild with the given copier and segment
+    /// map.
+    #[allow(dead_code)]
+    pub(super) fn new(map: RebuildMap, copier: T) -> Self {
+        let total_blks = map.count_dirty_blks();
+        let segment_size = map.segment_size();
+        let bit_vec: BitVec = map.into();
+        Self {
+            range: PeekableIterator::new(bit_vec.into_iter().enumerate()),
+            total_blks,
+            rebuilt_blks: 0,
+            segment_size,
+            copier: Rc::new(copier),
+        }
+    }
+}
+impl<T: RebuildTaskCopier> RangeRebuilder<T> for PartialRebuild<T> {
+    fn next(&mut self) -> Option<u64> {
+        for (blk, is_set) in self.range.by_ref() {
+            if is_set {
+                self.rebuilt_blks += self.segment_size;
```

Comment:
this should be `self.rebuilt_blks += self.segment_size_blks`

Reply:
I've renamed it to _blks